### PR TITLE
[WIP] Configure IdP and SP with configuration object

### PR DESCRIPTION
**Pull request by Zogoo**
The idea of this PR is that Gem should not care about how to store metadata. Because too much involving developer decisions makes this gem harder to maintain. 
https://github.com/saml-idp/saml_idp/issues/71

Let's drop persistance of metadata and how to refresh it because those implementation could be done with another gem (top of this gem) for different web frameworks. Or even it could be done by developers If we just accept "SamlIdp::IncomingMetadata" (IMO this should be also renamed)  for all configuration of IdP.

### CODE REVIEW

**Comment by jphenow** [comment]
Makes sense I think. Can much of the same be accomplished without the config? Might be good to document.

Would definitely say this is a change of API to the point that it might be good to place this in a new major version - in which case if we wanted to make a lot of breaking changes, might be good to figure out if we want a branch for vnext and maintain master for bugs until vnext is ready or how to go about managing the branches there.

**Comment by Zogoo** [comment]
> Makes sense I think. Can much of the same be accomplished without the config? Might be good to document.
> 
> Would definitely say this is a change of API to the point that it might be good to place this in a new major version - in which case if we wanted to make a lot of breaking changes, might be good to figure out if we want a branch for vnext and maintain master for bugs until vnext is ready or how to go about managing the branches there.

**Comment by Zogoo** [comment]
@jphenow, @mjobin-mdsol I have almost finished the required code changes and working on test cases. Could you guys please do a quick review to confirm the changes are okay?

The main change is that the `IdpConfig` and `SpConfig` objects allow to configure of the IdP for specific service providers. And it's not a global config because this allows 2 things
1. Gem users can use it for multiple SPs.
2. Gem users can make decisions about how they want to manage configs like DB, Global Object, Redis etc.

File: README.md [hunk 1]
```diff
@@ -20,186 +20,159 @@ Add this to your Gemfile:
...
+          name_id_format: SamlIdp::XML::Namespaces::Formats::NameId::EMAIL_ADDRESS
+        }
+      }
     end
+  end
+  
+  If you have a method called `asserted_attributes` in your Principal class,
+  there is no need to define it here in the config.
+
+  config.saml_attributes  =>
```
[the remark this answers is not in the document]
Comment (Zogoo) [reply README.md:130]: Oh, I forgot to mention that I haven't work documents yet. Once I finish code, I will update all documents.

**Review by mjobin-mdsol — COMMENTED** [review]
I glanced at it, but I am sorry I don't have useful comments at the moment.

**Comment by Zogoo** [comment]
This was too big PR for the current gem, I will create small PRs that allow transfer slowly to v2

**Pull request closed without merging**